chop/constraints.py

```diff
@@ -446,14 +446,111 @@ def prox(self, x, step_size=None):
     def __init__(self, a, b):
-        self.a = a 
+        """
+        Args:
+          a: float
+            min of the box constraint
+          b: float
+            max of the box constraint
+          """
+        if b < a:
+            raise ValueError(f"This constraint supposes that a <= b. Got {a}, {b}.")
+        self.a = a
         self.b = b
 
     def prox(self, x, step_size=None):
+        """Projection operator on the constraint.
+        Args:
+          x: torch.Tensor
+          step_size: Any
+
+        Returns:
+          x_thresh: torch.Tensor
+            x clamped between a and b.
+        """
         return torch.clamp(x, self.a, self.b)
+
+
+class Cone:
+    """
+    Represents second order cones of revolution centered in vector `u` (batch-wise), and angle :math: `\hat alpha`.
+    This constraint therefore really represents a batch of cones, which share the same half-angle.
+    The are all pointed in 0 (the origin).
+    Formally, the set is the following:
+
+    ..math::
+        \{x \in R^d,~ \|(uu^\top - Id)x\| \leq \alpha u^\top x \}
+
+    Note that :math: `\cos(\hat \alpha) = 1 / (1 + \alpha^2)`.
+    
+    The standard second order cone (ice-cream cone) is given by
+    `u = (0, ..., 0, 1)`, `cos_alpha=.5`.
+    """
+    def __init__(self, u, cos_angle=.05):
+        """
+        Args:
+          u: torch.Tensor
+            batch-wise directions centering the cones
+          cos_angle: float
+            cosine of the half-angle of the cone.
+        """
+        batch_size = u.size(0)
+        # normalize the cone directions
+        self.directions = utils.bmul(u, 1. / torch.norm(u.reshape(batch_size, -1), dim=-1))
+        self.cos_angle = cos_angle
+        self.alpha = np.sqrt(1. / cos_angle - 1)
+
+    def proj_u(self, x, step_size=None):
+        """
+        Projects x on self.directions batch-wise
+
+        Args:
+          x: torch.Tensor of shape (batch_size, *)
+            vectors to project
+          step_size: Any
+            Not used
+
+        Returns:
+          proj_x: torch.Tensor of shape (batch_size, *)
+            batch-wise projection of x onto self.directions
+        """
+        
+        return utils.bmul(utils.bdot(x, self.directions), self.directions)
+
+
+    @torch.no_grad()
+    def prox(self, x, step_size=None):
+        """
+        Projects `x` batch-wise onto the cone constraint.
+
+        Args:
+          x: torch.Tensor of shape (batch_size, *)
+            batch of vectors to project
+          step_size: Any
+            Not used
+
+        Returns:
+          proj_x: torch.Tensor of shape (batch_size, *)
+            batch-wise projection of `x` onto the cone constraint.
+        """
+        batch_size = x.size(0)
+        uTx = utils.bdot(self.directions, x)
+        p_u = self.proj_u(x)
+        p_orth_u = x - p_u
+        norm_p_orth_u = torch.norm(p_orth_u.reshape(batch_size, -1), dim=-1)
+        identity_idx = norm_p_orth_u <= self.alpha * uTx
```

Comment:
I would add a parenthesis for readability, i.e., ```identity_idx = (norm_p_orth_u <= self.alpha * uTx)``` . Same below and perhaps in other places